binderhub/builder.py

```diff
@@ -425,34 +424,75 @@ async def get(self, provider_prefix, _unescaped_spec):
-            api=self.settings.get("kubernetes_client"),
-            name=build_name,
-            namespace=self.settings["build_namespace"],
-            repo_url=repo_url,
-            ref=ref,
-            image_name=image_name,
-            push_secret=push_secret,
-            build_image=self.settings["build_image"],
-            memory_limit=self.settings["build_memory_limit"],
-            memory_request=self.settings["build_memory_request"],
-            docker_host=self.settings["build_docker_host"],
-            node_selector=self.settings["build_node_selector"],
-            appendix=appendix,
-            log_tail_lines=self.settings["log_tail_lines"],
-            git_credentials=provider.git_credentials,
-            sticky_builds=self.settings["sticky_builds"],
-        )
+        if issubclass(BuildClass, Build):
+            # Deprecated
+            build = BuildClass(
+                q=q,
+                # api object can be None if we are using FakeBuild
+                api=self.settings.get("kubernetes_client"),
+                name=build_name,
+                namespace=self.settings["build_namespace"],
+                repo_url=repo_url,
+                ref=ref,
+                image_name=image_name,
+                push_secret=push_secret,
+                build_image=self.settings["build_image"],
+                memory_limit=self.settings["build_memory_limit"],
+                memory_request=self.settings["build_memory_request"],
+                docker_host=self.settings["build_docker_host"],
+                node_selector=self.settings["build_node_selector"],
+                appendix=appendix,
+                log_tail_lines=self.settings["log_tail_lines"],
+                git_credentials=provider.git_credentials,
+                sticky_builds=self.settings["sticky_builds"],
+            )
+            build.main_loop = self.settings["ioloop"]
+        else:
+            build = BuildClass(
+                # Commented properties should be set in traitlets config
+                parent=self.settings["traitlets_parent"],
+                q=q,
+                name=build_name,
+                # namespace=self.settings["build_namespace"],
+                repo_url=repo_url,
+                ref=ref,
+                image_name=image_name,
+                # push_secret=push_secret,
+                # build_image=self.settings["build_image"],
+                # memory_limit=self.settings["build_memory_limit"],
+                # memory_request=self.settings["build_memory_request"],
+                # docker_host=self.settings["build_docker_host"],
+                # node_selector=self.settings["build_node_selector"],
+                appendix=appendix,
+                # log_tail_lines=self.settings["log_tail_lines"],
+                git_credentials=provider.git_credentials,
+                # sticky_builds=self.settings["sticky_builds"],
+                main_loop=self.settings["ioloop"],
```

Comment:
I made it an external parameter after some trial and error because I was getting errors due to `IOLoop.current()` not being set in the CI tests. I've changed it to be an internal parameter that's set in `BuildExecutor.__init__()` and the tests are passing. I tried defining it as a traitlet and setting it using
```python
    main_loop = Any(allow_none=False)

    @default("main_loop")
    def _default_main_loop(self):
        return IOLoop.current()
```
but that didn't work.

Reply:
> due to IOLoop.current() not being set in the CI tests

Making tests async should solve this, because there will always be a current loop if the loop is running. This also makes the tests reflect reality more accuractely, because when an app is running, there is always a current loop.

> I tried defining it as a traitlet and...that didn't work.

My guess is the default main loop traitlet didn't work because the first access could (at least sometimes) occur in the background thread, and the current loop is per-thread. I think setting it in init is the right way to go. 👍